src/main/java/au/org/ala/biocache/dao/SearchDAOImpl.java

```diff
@@ -863,15 +987,76 @@ public void writeCoordinatesToStream(SearchRequestParams searchParams,OutputStre
             StringBuilder infoHeader = new StringBuilder("infoHeaders");
             for (String h : header) infoHeader.append(",").append(h);
 
-            uidStats.put(infoFields.toString(), -1);
-            uidStats.put(infoHeader.toString(), -2);
+            uidStats.put(infoFields.toString(), new AtomicInteger(-1));
+            uidStats.put(infoHeader.toString(), new AtomicInteger(-2));
 
             //construct correct RecordWriter based on the supplied fileType
             final au.org.ala.biocache.RecordWriter rw = downloadParams.getFileType().equals("csv") ?
                     new CSVRecordWriter(out, header, downloadParams.getSep(), downloadParams.getEsc()) :
                     (downloadParams.getFileType().equals("tsv") ? new TSVRecordWriter(out, header) :
                             new ShapeFileRecordWriter(tmpShapefileDir, downloadParams.getFile(), out, (String[]) ArrayUtils.addAll(fields, qaFields)));
 
+            // Create a fixed length blocking queue for buffering results before they are written
+            final BlockingQueue<String[]> queue = new ArrayBlockingQueue<>(resultsQueueLength);
+            // Create a sentinel that we can check for reference equality to signal the end of the queue
+            final String[] sentinel = new String[0];
+            // An implementation of RecordWriter that adds to an in-memory queue
+            final RecordWriter concurrentWrapper = new RecordWriter() {
+                @Override
+                public void write(String[] nextLine) {
+                    try {
+                        if(Thread.currentThread().isInterrupted()) {
+                            finalise();
+                            return;
+                        }
+                        queue.offer(nextLine, writerTimeoutWaitMillis, TimeUnit.MILLISECONDS);
+                    } catch (InterruptedException e) {
```

Comment:
On any exceptions terminate outstanding callables. If possible flag the exception for the writer instead of waiting for the sentinel to be taken.